driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -405,30 +402,23 @@ public void onSet(Connection connection, Message.Response response, long latency
                                 return;
                             }
 
-                            logger.info("Query {} is not prepared on {}, preparing before retrying executing. "
-                                      + "Seeing this message a few times is fine, but seeing it a lot may be source of performance problems",
-                                        toPrepare.getQueryString(), connection.address);
                             String currentKeyspace = connection.keyspace();
                             String prepareKeyspace = toPrepare.getQueryKeyspace();
-                            // This shouldn't happen in normal use, because a user shouldn't try to execute
-                            // a prepared statement with the wrong keyspace set. However, if it does, we'd rather
-                            // prepare the query correctly and let the query executing return a meaningful error message
-                            if (prepareKeyspace != null && (currentKeyspace == null || !currentKeyspace.equals(prepareKeyspace)))
-                            {
-                                logger.debug("Setting keyspace for prepared query to {}", prepareKeyspace);
-                                connection.setKeyspace(prepareKeyspace);
+                            if (prepareKeyspace != null && (currentKeyspace == null || !currentKeyspace.equals(prepareKeyspace))) {
```

Comment:
If the session is not bound to any keyspace, then `currentKeyspace == null`, and the code will fail unless `prepareKeyspace == null`. That sounds like what we want, or am I missing something?